[MRG] DOC User guide section for histogram-based GBDTs (#14525)
scikit-learn/scikit-learn: doc/modules/ensemble.rst

======================

`Gradient Tree Boosting <https://en.wikipedia.org/wiki/Gradient_boosting>`_

> or Gradient Boosted Degression Trees (GBDT) is a generalization

Reviewer: For a second there I was wondering if "Degression Trees" were a thing that I should read about :)

Author: Oops ^^ thanks

of boosting to arbitrary
differentiable loss functions. GBDT is an accurate and effective
off-the-shelf procedure that can be used for both regression and
classification problems in a
variety of areas including Web search ranking and ecology.

The module :mod:`sklearn.ensemble` provides methods
for both classification and regression via gradient boosted decision
trees.

.. note::

  Scikit-learn 0.21 introduces two new experimental implementations of
  gradient boosting trees, namely :class:`HistGradientBoostingClassifier`
  and :class:`HistGradientBoostingRegressor`, inspired by
  `LightGBM <https://github.com/Microsoft/LightGBM>`_.

  These new histogram-based estimators can be **orders of magnitude faster**
  than :class:`GradientBoostingClassifier` and
  :class:`GradientBoostingRegressor` when the number of samples is larger
  than tens of thousands of samples.

  They also have built-in support for missing values, which avoids the need
  for an imputer. Support for categorical features is also part of the

>   roadmap.

Reviewer: I would not put "support of categorical features" in the user manual if it's not yet implemented.

Author: I thought it'd be a way to "engage" users?
We're not committing to any deadline so it's pretty safe. Open to removing though.

Reviewer: I think it's fine to talk about planned features in a blog post or a roadmap but not in a user manual. Otherwise if we documented each feature for which there is an open PR, the user manual would be very frustrating to read.

But let's see what other reviewers think.

Reviewer 2: Yeah the categorical will take more time, I'd leave it out. I'll put down the discussions we had with @ogrisel about it in an issue, and we can go forward from there.

  These estimators are described in more detail below in
  :ref:`histogram_based_gradient_boosting`.

  The following guide focuses on :class:`GradientBoostingClassifier` and
  :class:`GradientBoostingRegressor`, which might be preferred for small

>   sample sizes.

Reviewer: The previously explanation as to why they might be preferable sounds useful?

 * :ref:`sphx_glr_auto_examples_ensemble_plot_gradient_boosting_regression.py`

.. _histogram_based_gradient_boosting:

Histogram-Based Gradient Boosting
=================================

Scikit-learn 0.21 introduces two new experimental implementations of
gradient boosting trees, namely :class:`HistGradientBoostingClassifier`
and :class:`HistGradientBoostingRegressor`, inspired by
`LightGBM <https://github.com/Microsoft/LightGBM>`_.

> These new histogram-based estimators can be **orders of magnitude faster**

Reviewer: remove "new". Someone could read this 3-5 years from now in which case they will no longer be new :)

than :class:`GradientBoostingClassifier` and
:class:`GradientBoostingRegressor` when the number of samples is larger
than tens of thousands of samples.

They also have built-in support for missing values, which avoids the need
for an imputer. Support for categorical features is also part of the

> roadmap. Moreover, early-stopping is enabled by default.

Reviewer: Same comment about not mentioning "categorical features support" until it is implemented.

These fast estimators first bin the input samples ``X`` into
integer-valued bins (typically 256 bins) which tremendously reduces the

> number of splitting points to consider, and allow the algorithm to

Reviewer: ```suggestion
number of splitting points to consider, and allows the algorithm to
```

leverage integer-based data structures (histograms) instead of relying on

> sorted continuous values when building the trees. The API of these new

Reviewer: ```suggestion
sorted continuous values when building the trees. The API of these
```

estimators is slightly different, and some of the features from
:class:`GradientBoostingClassifier` and :class:`GradientBoostingRegressor`
are not yet supported: in particular sample weights, and some loss
functions.

> These new estimators are still **experimental** for now: their predictions

Reviewer: ```suggestion
These new estimators are still **experimental**: their predictions
```

and their API might change without any deprecation cycle. To use them, you
need to explicitly import ``enable_hist_gradient_boosting``::

  >>> # explicitly require this experimental feature
  >>> from sklearn.experimental import enable_hist_gradient_boosting  # noqa

>   >>> # now you can import normally from ensemble

Reviewer: This may be a bit redundant but fine.

  >>> from sklearn.ensemble import HistGradientBoostingClassifier

.. topic:: Examples:

 * :ref:`sphx_glr_auto_examples_inspection_plot_partial_dependence.py`

Usage
-----

Most of the parameters are unchanged from
:class:`GradientBoostingClassifier` and :class:`GradientBoostingRegressor`.

> One exception is the ``max_iter`` parameter that replaces ``n_estimators``.:

Reviewer: Maybe indicate that in both cases these parameters correspond to the number of iterations of the boosting process, as purely looking at the parameter name the equivalence is non trivial.

  >>> from sklearn.experimental import enable_hist_gradient_boosting
  >>> from sklearn.ensemble import HistGradientBoostingClassifier
  >>> from sklearn.datasets import make_hastie_10_2

  >>> X, y = make_hastie_10_2(random_state=0)
  >>> X_train, X_test = X[:2000], X[2000:]
  >>> y_train, y_test = y[:2000], y[2000:]
  >>> clf = HistGradientBoostingClassifier(max_iter=100).fit(X_train, y_train)

  >>> clf.score(X_test, y_test)
  0.8998

The size of the trees can be controlled through the ``max_lead_nodees``,
``max_depth``, and ``min_samples_leaf`` parameters.

The number of bins used to bin the data is controlled with the ``max_bins``

> parameter. Using less bins acts as some sort of regularization. It is

Reviewer: ```suggestion
parameter. Using less bins acts as a form of regularization. It is
```

> generally recommended to use as many bins as possible.

Reviewer: Maybe could be reformulated. How many are possible? I imagine if I set `max_bins=int(1e9)` that would also not be good..

Author: max_bins can't be more than 255 anyway, I'll make it clearer

The ``l2_regularization`` parameter is a regularizer on the loss function and
corresponds to :math:`\lambda` in equation (2) of [XGBoost]_.

> Note that unlike most estimators, **early-stopping is enabled by default**.

Reviewer: ```suggestion
Note that **early-stopping is enabled by default**.
```

I think there is a discussion on changing this for other estimators as well.

training or validation loss.

Missing values support
----------------------

:class:`HistGradientBoostingClassifier` and
:class:`HistGradientBoostingRegressor` have built-in support for missing
values (NaNs).

During training, the tree grower learns at each split point whether samples
with missing values should go to the left or right child, based on the
potential gain. When predicting, samples with missing values are assigned to
the left or right child consequently.

.. TODO: Add this example when missing values PR is merged (results are
.. wrong for now)

.. from sklearn.experimental import enable_hist_gradient_boosting
.. from sklearn.ensemble import HistGradientBoostingRegressor
.. import numpy as np

.. X = np.array([0, 1, 2, np.nan]).reshape(-1, 1)
.. y = [0, 0, 1, 1]
.. gbdt = HistGradientBoostingRegressor().fit(X, y)
.. gbdt.predict(X)

If no missing values were encountered for a given feature during training,
then samples with missing values are mapped to whichever child has the most
samples.

Low-level parallelism
---------------------

:class:`HistGradientBoostingClassifier` and
:class:`HistGradientBoostingRegressor` have parallel implementations that

> use OpenMP through Cython. The number of threads that is used can be changed

Reviewer: Maybe add that by default all available threads are used?

using the ``OMP_NUM_THREADS`` environment variable. Please refer to the
OpenMP documentation for details. We are planning on adding a ``n_jobs``

> parameter (or equivalent) in a future version.

Reviewer: Maybe worth removing last sentence until it is done.

 * :ref:`sphx_glr_auto_examples_ensemble_plot_gradient_boosting_regression.py`

.. _histogram_based_gradient_boosting:

Histogram-Based Gradient Boosting
=================================

Scikit-learn 0.21 introduces two new experimental implementations of
gradient boosting trees, namely :class:`HistGradientBoostingClassifier`
and :class:`HistGradientBoostingRegressor`, inspired by
`LightGBM <https://github.com/Microsoft/LightGBM>`_.

These histogram-based estimators can be **orders of magnitude faster**
than :class:`GradientBoostingClassifier` and
:class:`GradientBoostingRegressor` when the number of samples is larger
than tens of thousands of samples.

They also have built-in support for missing values, which avoids the need
for an imputer.

These fast estimators first bin the input samples ``X`` into

> integer-valued bins (typically 256 bins) which tremendously reduces the

Reviewer 3: Will a user get confused here with `max_bins`?

Author: It's is detailed that one bin is reserved for missing values in the docstring.

Reviewer 3: Later in this user guide it says:

> It is generally recommended to use as many bins as possible, which is the default: 255 bins.

which can be confusing.

Even over here we are binning `X` with 255 bins, since we are always saving one bind for missing.

Author: I removed the 255

  >>> from sklearn.ensemble import HistGradientBoostingClassifier
  >>> from sklearn.datasets import make_hastie_10_2

  >>> X, y = make_hastie_10_2(random_state=0)
  >>> X_train, X_test = X[:2000], X[2000:]
  >>> y_train, y_test = y[:2000], y[2000:]
  >>> clf = HistGradientBoostingClassifier(max_iter=100).fit(X_train, y_train)

  >>> clf.score(X_test, y_test)
  0.8998

The size of the trees can be controlled through the ``max_leaf_nodes``,
``max_depth``, and ``min_samples_leaf`` parameters.

The number of bins used to bin the data is controlled with the ``max_bins``
parameter. Using less bins acts as a form of regularization. It is
generally recommended to use as many bins as possible, which is the default:
255 bins.

The ``l2_regularization`` parameter is a regularizer on the loss function and
corresponds to :math:`\lambda` in equation (2) of [XGBoost]_.

Note that **early-stopping is enabled by default**. The early-stopping
behaviour is controlled via the ``scoring``, ``validation_fraction``,
``n_iter_no_change``, and ``tol`` parameters. It is possible to early-stop
using an arbitrary :term:`scorer`, or just the training or validation loss. By

> default, early-stopping is performed using the the default :term:`scorer` of

Author: ```suggestion
default, early-stopping is performed using the default :term:`scorer` of
```

During training, the tree grower learns at each split point whether samples
with missing values should go to the left or right child, based on the
potential gain. When predicting, samples with missing values are assigned to
the left or right child consequently.

.. TODO: Add this example when missing values PR is merged (results are
.. wrong for now)

.. from sklearn.experimental import enable_hist_gradient_boosting
.. from sklearn.ensemble import HistGradientBoostingRegressor
.. import numpy as np

.. X = np.array([0, 1, 2, np.nan]).reshape(-1, 1)
.. y = [0, 0, 1, 1]
.. gbdt = HistGradientBoostingRegressor().fit(X, y)
.. gbdt.predict(X)

If no missing values were encountered for a given feature during training,
then samples with missing values are mapped to whichever child has the most
samples.

Low-level parallelism
---------------------

:class:`HistGradientBoostingClassifier` and
:class:`HistGradientBoostingRegressor` have implementations that use OpenMP
for parallelization through Cython. The number of threads that is used can
be changed using the ``OMP_NUM_THREADS`` environment variable. By default,
all available cores are used. Please refer to the OpenMP documentation for

> details.

Reviewer 3: Link to OpenMP documentation?

Author: (Weird that the previous comments got lost, we discussed this with Adrin already)

I don't want to add it because the openmp doc is a 300 pages pdf. Googling OMP_NUM_THREADS gives you the docs from intel and ibm but they're probably not official.

Why it's faster
---------------

The bottleneck of a gradient boosting procedure is building the decision
trees. Building a traditional decision tree (as in the other GBDTs
:class:`GradientBoostingClassifier` and :class:`GradientBoostingRegressor`)

> requires sorting the values of all the samples at each node (for

Author: ```suggestion
requires sorting the samples at each node (for
```

> requires sorting the feature values of all the samples at each node (for

Reviewer 2: sorry, left the comment on the commit, `the feature values of all` kinda sounds redundant here.

> requires sorting the feature values of all the samples at each node (for

Author: ```suggestion
requires sorting the values of all the samples at each node (for
```

each feature). Sorting is needed so that the potential gain of a split point
can be computed efficiently. Splitting a single node has thus a complexity
of :math:`\mathcal{O}(\text{n_features} * n \log(n))` where :math:`n` is the
number of samples at the node.

:class:`HistGradientBoostingClassifier` and
:class:`HistGradientBoostingRegressor`, in contrast, do not require sorting the
feature values and instead use a data-structure called a histogram, where the
samples are implicitly ordered. Building a histogram has a
:math:`\mathcal{O}(n)` complexity, so the node splitting procedure has a
:math:`\mathcal{O}(\text{n_features} * n)` complexity, much smaller than the
previous one. In addition, instead of considering :math:`n` split points, we
here consider only ``max_bins`` split points, which is much smaller.

In order to build histograms, the input data `X` needs to be binned into
integer-valued bins. This binning procedure does require sorting the feature
values, but it only happens once at the very beginning of the boosting process
(not at each node, like in :class:`GradientBoostingClassifier` and
:class:`GradientBoostingRegressor`).

> Finally, many parts of the implementation of

Reviewer 3: There are four places, we can list them?

1. Binning
2. Building Histograms
3. Splitting
3. Predicting

Author: good idea

parameter. Using less bins acts as a form of regularization. It is
generally recommended to use as many bins as possible, which is the default.

The ``l2_regularization`` parameter is a regularizer on the loss function and
corresponds to :math:`\lambda` in equation (2) of [XGBoost]_.

Note that **early-stopping is enabled by default**. The early-stopping
behaviour is controlled via the ``scoring``, ``validation_fraction``,
``n_iter_no_change``, and ``tol`` parameters. It is possible to early-stop
using an arbitrary :term:`scorer`, or just the training or validation loss. By
default, early-stopping is performed using the default :term:`scorer` of
the estimator on a validation set.

Missing values support
----------------------

:class:`HistGradientBoostingClassifier` and
:class:`HistGradientBoostingRegressor` have built-in support for missing
values (NaNs).

During training, the tree grower learns at each split point whether samples
with missing values should go to the left or right child, based on the
potential gain. When predicting, samples with missing values are assigned to
the left or right child consequently.

> .. TODO: Add this example when missing values PR is merged (results are

Reviewer 2: Do you want to mention the missing values in that PR then? I mean, remove it from here, and add it to this doc in that PR.

Author: Let's leave it here? it's just a comment it is not rendered. Moving it to the missing values PR requires merging this one first anyway

Reviewer 2: I meant the whole "missing values" section, not the example.

Author: ok done

 * :ref:`sphx_glr_auto_examples_ensemble_plot_gradient_boosting_regression.py`

.. _histogram_based_gradient_boosting:

Histogram-Based Gradient Boosting
=================================

> Scikit-learn 0.21 introduces two new experimental implementations of

Reviewer 2: If we say they're experimental here, we need to have a marker to make sure we remember to fix it when we move it out of being experimental.

Author: hard to do since we don't have a target version. Making an estimator not experimental won't be trivial (lots of files will be changed) so I'm pretty sure we won't miss the user guide

and their API might change without any deprecation cycle. To use them, you
need to explicitly import ``enable_hist_gradient_boosting``::

  >>> # explicitly require this experimental feature
  >>> from sklearn.experimental import enable_hist_gradient_boosting  # noqa
  >>> # now you can import normally from ensemble
  >>> from sklearn.ensemble import HistGradientBoostingClassifier

.. topic:: Examples:

 * :ref:`sphx_glr_auto_examples_inspection_plot_partial_dependence.py`

Usage
-----

Most of the parameters are unchanged from
:class:`GradientBoostingClassifier` and :class:`GradientBoostingRegressor`.
One exception is the ``max_iter`` parameter that replaces ``n_estimators``, and
controls the number of iterations of the boosting process:

  >>> from sklearn.experimental import enable_hist_gradient_boosting
  >>> from sklearn.ensemble import HistGradientBoostingClassifier
  >>> from sklearn.datasets import make_hastie_10_2

  >>> X, y = make_hastie_10_2(random_state=0)
  >>> X_train, X_test = X[:2000], X[2000:]
  >>> y_train, y_test = y[:2000], y[2000:]
  >>> clf = HistGradientBoostingClassifier(max_iter=100).fit(X_train, y_train)

  >>> clf.score(X_test, y_test)
  0.8998

> The size of the trees can be controlled through the ``max_lead_nodees``,

Author: ```suggestion
The size of the trees can be controlled through the ``max_leaf_nodes``,
```

``max_depth``, and ``min_samples_leaf`` parameters.

The number of bins used to bin the data is controlled with the ``max_bins``
parameter. Using less bins acts as a form of regularization. It is
generally recommended to use as many bins as possible, which is the default:

> 255 bins for non-missing values.

Reviewer 2: I think we either should leave the "non-missing" value out, or explain what you mean here.

The ``l2_regularization`` parameter is a regularizer on the loss function and

> corresponds to :math:`\lambda` in equation (2) of [XGBoost]_.

Reviewer 2: should it be 

    `XGBoost`_

instead?

Author: We use the []_ syntax for paper references as far as I know

default, early-stopping is performed using the the default :term:`scorer` of
the estimator on a validation set.

Missing values support
----------------------

:class:`HistGradientBoostingClassifier` and
:class:`HistGradientBoostingRegressor` have built-in support for missing
values (NaNs).

During training, the tree grower learns at each split point whether samples
with missing values should go to the left or right child, based on the
potential gain. When predicting, samples with missing values are assigned to
the left or right child consequently.

.. TODO: Add this example when missing values PR is merged (results are
.. wrong for now)

.. from sklearn.experimental import enable_hist_gradient_boosting
.. from sklearn.ensemble import HistGradientBoostingRegressor
.. import numpy as np

.. X = np.array([0, 1, 2, np.nan]).reshape(-1, 1)
.. y = [0, 0, 1, 1]
.. gbdt = HistGradientBoostingRegressor().fit(X, y)
.. gbdt.predict(X)

If no missing values were encountered for a given feature during training,
then samples with missing values are mapped to whichever child has the most
samples.

Low-level parallelism
---------------------

:class:`HistGradientBoostingClassifier` and

> :class:`HistGradientBoostingRegressor` have parallel implementations that

Reviewer 2: maybe "have implementations that use OpenMP for parallelization through Cython".

use OpenMP through Cython. The number of threads that is used can be changed
using the ``OMP_NUM_THREADS`` environment variable. By default, all available

> cores are used. Please refer to the OpenMP documentation for details. We are

Reviewer 2: maybe a link to openmp's docs?

Author: it's tricky, the openmp doc is a huge pdf. I think users are better off just googling `openmp OMP_NUM_THREADS` (first links are from intel and ibm but since these aren't official I'd rather not link to them?)
